### DISCUSSION TITLE
Sending exception to Sentry

### DISCUSSION THREAD

**Discussion by abkfenris** [dagster-io/dagster#11102] [Ideas]
While Dagster's logging is really nice, it doesn't always give you quite as much information right when an exception happens as some other tools. Personally I use [Sentry.io](https://sentry.io/) to capture exceptions in other things I build, so I tried to get it integrated into Dagster jobs.

The main way that data gets sent to Sentry is by attaching to [`sys.excepthook`](https://docs.python.org/3/library/sys.html#sys.excepthook). Dagster however [wraps all user functions with a broad exception handler](https://docs.dagster.io/_apidocs/internals#dagster._core.errors.user_code_error_boundary), so exceptions never make it up to the system level. Sentry however will pick up that something bad has happened via log messages, but it won't have any of the context that usually makes it so useful.

It turns out that two can play the game of intercepting exceptions, so here's how I've gotten Sentry and Dagster to play nice.

First I made a helper to set up the Sentry client without a few of the usual integrations, and get it to ignore the `dagster` logger.
```py
def setup_dagster_sentry():
    """
    Setup the sentry SDK for Dagster if SENTRY_DSN is defined for the environment.

    Additionally if IMAGE_TAG is defined it will be set as the release,
    and TRACES_SAMPLE_RATE can be set 0-1 otherwise will default to 0.

    Manually sets up a bunch of the default integrations and disables logging of dagster
    to quiet things down.
    """
    from sentry_sdk.integrations.argv import ArgvIntegration
    from sentry_sdk.integrations.atexit import AtexitIntegration
    from sentry_sdk.integrations.dedupe import DedupeIntegration
    from sentry_sdk.integrations.logging import LoggingIntegration, ignore_logger
    from sentry_sdk.integrations.modules import ModulesIntegration
    from sentry_sdk.integrations.stdlib import StdlibIntegration

    ignore_logger("dagster")

    SENTRY_DSN = os.environ.get("SENTRY_DSN")
    IMAGE_TAG = os.environ.get("IMAGE_TAG")
    TRACES_SAMPLE_RATE = float(os.environ.get("SENTRY_TRACES_SAMPLE_RATE", 0))

    if SENTRY_DSN:
        sentry_sdk.init(
            dsn=SENTRY_DSN,
            traces_sample_rate=TRACES_SAMPLE_RATE,
            release=IMAGE_TAG,
            default_integrations=False,
            integrations=[
                AtexitIntegration(),
                DedupeIntegration(),
                StdlibIntegration(),
                ModulesIntegration(),
                ArgvIntegration(),
                LoggingIntegration(),
            ],
        )
```
Then I created my own decorator to sit in between `dagster.op` send exceptions to Sentry before letting Dagster at them.

It also emits the unique Sentry event ID into the Dagster logs to help in finding the error.
```py
def log_op_context(context: "OpExecutionContext"):
    """
    Capture Dagster OP context for Sentry Error handling
    """
    sentry_sdk.add_breadcrumb(
        category="dagster",
        message=f"{context.job_name} - {context.op_def.name}",
        level="info",
        data={
            "run_config": context.run_config,
            "job_name": context.job_name,
            "op_name": context.op_def.name,
            "run_id": context.run_id,
            "retry_number": context.retry_number,
        },
    )

    sentry_sdk.set_tag("job_name", context.job_name)
    sentry_sdk.set_tag("op_name", context.op_def.name)
    sentry_sdk.set_tag("run_id", context.run_id)

def capture_op_exceptions(func):
    """
    Captures exceptions thrown by Dagster Ops and forwards them to Sentry
    before re-throwing them for Dagster.

    Expects ops to receive Dagster context as the first argument,
    but it will continue if it doesn't (it just won't get as much context).

    It will log a unique ID that can be then entered into Sentry to find
    the exception.

    This should be used as a decorator between Dagster's `@op`,
    and the function to be handled.

    @op
    @sentry.capture_op_exceptions
    def op_with_error(context):
        raise Exception("Ahh!")
    """
    from dagster import get_dagster_logger

    @functools.wraps(func)
    def wrapped_fn(*args, **kwargs):
        logger = get_dagster_logger("sentry")

        try:
            log_op_context(args[0])
        except (AttributeError, IndexError):
            logger.warn("Sentry did not find execution context as the first arg")

        try:
            return func(*args, **kwargs)
        except Exception as e:
            event_id = sentry_sdk.capture_exception(e)
            logger.error(f"Sentry captured an exception. Event ID: {event_id}")
            raise e

    return wrapped_fn
```

Then it's used as a decorator on the ops.

```py
@op
@capture_op_exception
def might_explode(context):
    raise Exception("Oops!")
```

Dagster now lets us know that a Sentry error has been made.

<img width="544" alt="image" src="https://user-images.githubusercontent.com/1296209/207462997-72286138-73ed-446f-af9d-d1c068fdc430.png">

And Sentry has all sorts of context of what went wrong.

<img width="1084" alt="image" src="https://user-images.githubusercontent.com/1296209/207463087-a43615f6-f426-4dfb-9eba-92744f7e4dbc.png">

There are probably other ways that Sentry could be closer integrated (such as using `dagster.logger` with the various Sentry logging handlers), but now between Sentry and Dagster I have enough context to deal with many more errors without having to log every variable along the way.

**Comment by EtienneT** [comment]
Thank you @abkfenris for the amazing solution, works great.  I was wondering if you figured out a way to avoid having to add `@capture_op_exception` to all your ops and assets manually?

For example I tried using capture_op_exception as a simple function wrapper instead of a decorator.  This would allow us to wrap all assets with capture_op_exception before inserting them to the Definitions object.  So we could do:

```
all_assets = [capture_asset_exceptions(x) for x in all_assets]

def = Definitions(assets = all_assets, ...)
```
Unfortunately we get an error related to type checking:
```
dagster._check.CheckError: Member of iterable mismatches type. Expected (<class 'dagster._core.definitions.assets.AssetsDefinition'>,
<class 'dagster._core.definitions.source_asset.SourceAsset'>, <class 'dagster._core.definitions.cacheable_assets.CacheableAssetsDefinition'>).
Got <function capture_asset_exceptions.<locals>.wrapped_fn at 0x000002604375FD80> of type <class 'function'>.
```
Looking at the code for `@asset`, I can't see an easy fix.

@capture_op_exception is also sensitive to where you insert it.  You have to insert it after `@asset` to be able to add the asset in the Definitions object, otherwise you get the same error as above.

It would be great to have an official Sentry integration given Sentry's popularity.

Thanks again for the code!

**Reply by abkfenris** [reply]
Yes, it does have to between `@op` or `@asset` to get the context, and to make sure that Dagster can figure out when the right time to run it is.

If you don't want to manage remembering to sneak it in between, I'd probably make a decorator that combines `@asset` and `@capture_op_exceptions`, and takes all the kwargs of `@asset` and passes them along, then use that in place of `@asset`.

**Comment by whalesalad** [comment]
Want to bump this because this is something we would love to see. Our plan is to fork Dagster and submit a PR to integrate this, but we do not have rich understanding of how the system works currently. The investigation on this page though is priceless thank you for posting.

Would be great to do something like

```
import dagster

dagster.add_exception_handler('sentry', fn)
```

So that any number of these handlers can be added. The exception would get broadcast to all subscribed handlers, including the build-in one. 

The caveat is that exceptions are heavy and in this case we don't want to _handle_ them (in a blocking manner) as much as we just want to get notified of them and get them as data to pass it on elsewhere.

So perhaps `dagster.on_exception` is a better name.

Anyway, we've been huge fans of Dagster and have a lot of our workloads ported over to it, but atm we essentially need to manually visit the UI periodically to check in on failures. We'll embark on trying to do this ourselves unless the Dagster team has a better design idea or a quicker way to tackle.

**Comment by davidemerritt** [comment]
+1 - If the dagster team can help enable this, it would be a huge help to our organization. We'd love to have full exception tracking/tracing via sentry.

**Comment by ydennisy** [comment]
+1 on this, it would be a great integration to have!

**Comment by lautarodibartolo-ae** [comment]
+1 to get this in Dagster!

**Comment by zachoooo** [comment]
Would love to see how this is meant to be done. I also don't want to have to wrap all of my assets like this and would love to have a way to initialize in a way that ensures that all asset executions have an exception handling context from Sentry or Bugsnag.

**Comment by ddelange** [comment]
A simple (global) sentry integration by registering a [`dagster.failure_hook`](https://release-1-11-14.archive.dagster-docs.io/api/dagster/hooks#dagster.failure_hook):

Any feedback is appreciated! Would love to see a version of this land as something like `dagster.register_sentry()` ❤️ 

```py
import os
import sentry_sdk
import dagster as dg


if os.environ.get("SENTRY_DSN"):
    sentry_sdk.init(
        dsn=os.environ.get("SENTRY_DSN"),
        environment=os.environ.get("SENTRY_ENVIRONMENT"),
        release=os.environ.get("SENTRY_RELEASE"),
    )


def send_sentry_issue(*, exception, job_name, run_id, step_key, op_name):
    """Send exception to Sentry with context and aggregation by exception type + stacktrace ref https://docs.sentry.io/concepts/data-management/event-grouping/"""
    with sentry_sdk.push_scope() as scope:
        # This is the default but making it explicit
        scope.level = "error"
        # Set structured context for rich event data
        scope.set_context(
            "dagster",
            {
                "job_name": job_name,
                "run_id": run_id,
                "step_key": step_key,
                "op_name": op_name,
            },
        )
        # Add breadcrumb for execution trace
        sentry_sdk.add_breadcrumb(
            category="dagster",
            message=f"Op '{op_name}' failed in job '{job_name}'",
            level="error",
            data={
                "step_key": step_key,
                "run_id": run_id,
            },
        )
        # Send to sentry
        sentry_sdk.capture_exception(exception)


@dg.failure_hook
def sentry_on_failure(context: dg.HookContext):
    """Hook that sends exceptions to Sentry using HookContext ref https://docs.dagster.io/api/dagster/hooks#dagster.HookContext"""
    if sentry_sdk.is_initialized():
        send_sentry_issue(
            exception=context.op_exception,
            job_name=context.job_name,
            run_id=context.run_id,
            step_key=context.step_key,
            op_name=context.op.name,
        )
```

**Comment by cbility** [comment]
+1, wondering if native support for this is on the roadmap?

**Comment by 1oglop1** [comment]
This is a must for every company taking software seriously.
